### PR TITLE
Add error codes for libsyntax_ext

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
I intend to add error codes for `libsyntax_ext` as well. However, they cannot be used at stage 0 directly so I thought it might be possible to enable them at the stage 1 only so we can have access to the macros. However, the error code registration seems to not work this way. Currently I get the following error:

```
error: used diagnostic code E0660 not registered
  --> libsyntax_ext/asm.rs:93:25
   |
93 |                         span_err!(cx, sp, E0660, "malformed inline assembly");
   |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error: used diagnostic code E0661 not registered
   --> libsyntax_ext/asm.rs:151:33
    |
151 | /                                 span_err!(cx, sp, E0661,
152 | |                                           "output operand constraint lacks '=' or '+'");
    | |________________________________________________________________________________________^
    |
    = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error: aborting due to 2 previous errors

error: Could not compile `syntax_ext`.
```

If anyone has an idea, I'd gladly take it. I'm trying to figure this out on my side as well. I also opened this PR to know if it was worth it to continue (maybe we don't want this?).

Anyway, any answer for both questions is very welcome!

cc @rust-lang/compiler

### CODE REVIEW

**Comment by pietroalbini** [comment]
@rust-lang/compiler, can we get someone to answer @GuillaumeGomez questions?

**Comment by nikomatsakis** [comment]
I have no idea what the problem is :(

**Comment by petrochenkov** [comment]
I personally have dislike to error codes in general.

**Comment by GuillaumeGomez** [comment]
That helps. 😆

**Comment by nikomatsakis** [comment]
@GuillaumeGomez where is your branch? did you opt into the feature gate?

**Comment by nikomatsakis** [comment]
oh, @eddyb also says that you need to do `#[cfg(stage0)]` until beta is fixed... so maybe try again when the next beta lands...?

**Comment by nikomatsakis** [comment]
see also: 

https://github.com/rust-lang/rust/blob/b1f8e6fb06d7362eeb2065347a7db94e76b1cb2f/src/librustc/lib.rs#L182

**Comment by GuillaumeGomez** [comment]
I'll wait for next beta then.

**Comment by pietroalbini** [comment]
@GuillaumeGomez I think a new beta was released by now?

**Comment by GuillaumeGomez** [comment]
Cool, I'll come back to it then.

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by GuillaumeGomez** [comment]
Ah, right, the tests I added can only work on stage 2 compiler... Is there a way to make these two tests to be run only on stage 2?

**Comment by shepmaster** [comment]
This never got a reviewer...

r? @eddyb

**Comment by nikomatsakis** [comment]
Travis is unhappy, to start:

https://travis-ci.org/rust-lang/rust/builds/358583412#L1748

**Comment by GuillaumeGomez** [comment]
@nikomatsakis: yes because tests can only be run on stage2. Which is why I asked:

> Ah, right, the tests I added can only work on stage 2 compiler... Is there a way to make these two tests to be run only on stage 2?

**Comment by eddyb** [comment]
@GuillaumeGomez I think it's `// ignore-stage1` (if you search around, there should be examples).

File: src/libsyntax_ext/diagnostics.rs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+// except according to those terms.
+
+#![allow(non_snake_case)]
+
+// Error messages for EXXXX errors.
+// Each message should start and end with a new line, and be wrapped to 80 characters.
+// In vim you can `:set tw=80` and use `gq` to wrap paragraphs. Use `:set tw=0` to disable.
+register_long_diagnostics! {
+E0660: r##"
+An invalid syntax was passed to the `asm` macro.
```
Comment (nikomatsakis) [inline src/libsyntax_ext/diagnostics.rs]: Grammatical nit: I don't think of "syntax" as something you can have "one" of. I would rather say, "The argument to the `asm` macro is not well-formed" or something like that, perhaps?

Comment (GuillaumeGomez) [reply src/libsyntax_ext/diagnostics.rs]: Fine by me!

**Comment by GuillaumeGomez** [comment]
Fixed tests.

**Comment by RalfJung** [comment]
> However, they cannot be used at stage 0 directly 

Care to explain why? Was this just a problem with the old bootstrap compiler?

This stage-0-conditional is still present in current master, and causing confusion like https://github.com/rust-lang/rust/issues/53380#issuecomment-413160669

**Comment by GuillaumeGomez** [comment]
We need some macros (which are defined later on) to be defined.

**Comment by RalfJung** [comment]
But why are those not in the beta bootstrap compiler?

**Comment by GuillaumeGomez** [comment]
Hard to explain since I don't have the precise context in mind. Just gave you what I had in memory. ;)

**Pull request merged**